cryptofeed/exchanges/binance.py

```diff
@@ -288,36 +326,136 @@ async def _snapshot(self, pair: str) -> None:
+                    self._l2_book[std_pair].book[side][price] = amount
+                    if i < max_depth:
+                        delta[side].append((price, amount))
+        self._l2_book[std_pair].sequence_number = resp['lastUpdateId']
+        
+        if self.refresh_snapshot:
+            # Delete buffer and reset time since snapshot update
+            self._next_snapshot_time[std_pair] = receipt_timestamp + self._next_snapshot_refresh_interval()
+            self._fetch_snapshop_buffer.pop(std_pair, None)
+
+        await self.book_callback(L2_BOOK, self._l2_book[std_pair], receipt_timestamp, raw=resp, delta=delta, sequence_number=self.last_update_id[std_pair])
+
+    async def _refresh_snapshot_with_buffer(self, pair: str) -> None:
+        resp = await self._fetch_snapshot(pair)
+        receipt_timestamp = time.time()
+
+        std_pair = self.exchange_symbol_to_std_symbol(pair)
+        if std_pair in self._book_buffer or std_pair not in self._fetch_snapshop_buffer:
+            # already fetching / fetched a snapshot, so ignore this message
+            self._fetch_snapshop_buffer.pop(std_pair, None)
             return
 
+        while resp['lastUpdateId'] > self.last_update_id[std_pair]:
+            # Snapshot came before we received the corresponding WS update. Wait until we receive that.
+            if self.depth_interval == '100ms':
+                await sleep(0.1)
+            elif self.depth_interval == '1000ms':
+                await sleep(0.5)
+
+        # Construct a temporary book from snapshot
+        temp_book = OrderBook(self.id, std_pair, max_depth=self.max_depth, bids={Decimal(u[0]): Decimal(u[1]) for u in resp['bids']}, asks={Decimal(u[0]): Decimal(u[1]) for u in resp['asks']})
+
+        # Apply buffered messages to temporary book
+        while len(self._fetch_snapshop_buffer[std_pair]) > 0:
+            msg = self._fetch_snapshop_buffer[std_pair].popleft()
+            self._apply_msg_to_book(temp_book.book, msg)
+        
+        # Apply temporary book to local l2 book and resolve deltas
+        max_depth = self.max_depth if self.max_depth else self.valid_depths[-1]
+        delta = {BID: [], ASK: []}
+        for side in [BID, ASK]:
+            for i, price in enumerate(temp_book.book[side]):
+                amount = temp_book.book[side][price]
+                if price in self._l2_book[std_pair].book[side]:
+                    if amount != self._l2_book[std_pair].book[side][price]:
+                        # This should never happen
+                        self._reset()
+                        LOG.error("%s: Book amount not equal to snapshot amount after fetching a refresh snapshot, resetting book", self.id)
+                        return
+                else:
+                    # Price not in order book: add it there
+                    self._l2_book[std_pair].book[side][price] = amount
+                    if i < max_depth:
+                        delta[side].append((price, amount))
```

Comment:
not sure if this is safe for users who omit `self.max_depth` and want the internal book to be as large & correct as possible

I havent tested this but pretty sure that `self.max_depth` being set will enforce the book to filter before sending callback deltas anyway? I'm thinking this should always be adding a delta...

Reply:
You're right, limiting deltas to max_depth is a problem if you want the book to be as big as possible. However, `max_depth` does not filter the deltas being sent to downstream consumers. That means we do need to do this filtering somewhere, otherwise a downstream consumer with max_depth=20 would see deltas for depth=38 for instance (because we fetch a snapshot of 100 depths minimum). In fact this problem exists without the snapshots as well, because the web socket could get a book update orders higher on the order book.

I'm not immediately sure how to fix this in the most elegant way. You could either change the cause to be `if i < max_depth or not self.max_depth` but that doesn't feel very nice and only fixes the issue for snapshots, not for book updates. Alternatively you could filter the deltas in the callback if max_depth is set, but that's very inefficient and rather cumbersome. @bmoscon what do you think?

Reply:
Just to show that `max_depth` is not filtering deltas: when running the demo on `master` with BTC-BUSD, I got the following update:
```
Snap at 1632908947.502313 (truncated book from 20 to 20):
42221.54|0.451, 42221.57|0.021, 42221.58|0.176, 42221.65|0.083, 42223.44|0.062, 42223.45|0.083, 42223.50|0.035, 42224.11|0.059, 42225.45|0.011, 42227.01|0.001, 42228.37|0.118, 42230.47|0.128, 42230.48|0.010, 42230.49|0.054, 42230.57|0.250, 42230.98|0.002, 42231.41|0.128, 42231.68|0.059, 42232.40|0.050, 42232.78|0.126
Deltas:	[(Decimal('42221.54000000'), Decimal('0.45091000')), (Decimal('42221.57000000'), Decimal('0.02130000')), (Decimal('42221.58000000'), Decimal('0.17597000')), (Decimal('42221.61000000'), Decimal('0E-8')), (Decimal('42234.73000000'), Decimal('0.22972000')), (Decimal('42258.50000000'), Decimal('4.05378000')), (Decimal('42305.55000000'), Decimal('0.21000000'))]
```
As you can see, the last few deltas fall outside our range. That's because the Binance web socket always sends all book updates to us, even if they were outside our requested `max_depth`, and the code that filters deltas was removed some time ago.

Reply:
thats expected. at some point ill add delta filtering either in the cython module or in the c module. its far too slow to try and do it on the python side. so, I'd say this is expected behavior for the time being